pilot/pkg/config/cloudfoundry/controller.go

```diff
@@ -0,0 +1,320 @@
+	client        CopilotClient
+	store         model.ConfigStore
+	logger        logger
+	timeout       time.Duration
+	checkInterval time.Duration
+	eventHandlers []func(model.Config, model.Event)
+	storage       storage
+}
+
+// NewController provides a new CF controller
+func NewController(client CopilotClient, store model.ConfigStore, logger logger, timeout, checkInterval time.Duration) *controller {
+	return &controller{
+		client:        client,
+		store:         store,
+		logger:        logger,
+		timeout:       timeout,
+		checkInterval: checkInterval,
+		storage: storage{
+			virtualServices:  make(map[string]*model.Config),
+			destinationRules: make(map[string]*model.Config),
+		},
+	}
+}
+
+// RegisterEventHandler will save event handler off to controller struct
+// we call handler on a time interval instead within Run
+func (c *controller) RegisterEventHandler(typ string, f func(model.Config, model.Event)) {
+	c.eventHandlers = append(c.eventHandlers, f)
+}
+
+// HasSynced demonstrates whether we have route data or not - meaning
+// we've successfuclly queried the copilot for data
+func (c *controller) HasSynced() bool {
+	return len(c.storage.virtualServices) > 0 && len(c.storage.destinationRules) > 0
+}
+
+// Run kicks off the async copilot fetch plus running
+// the cache clearing event handler that causes the rest of
+// pilot to repopulate the rules to push data to envoy
+func (c *controller) Run(stop <-chan struct{}) {
+	tick := time.NewTicker(c.checkInterval)
+
+	go func() {
+		for {
+			select {
+			case <-stop:
+				tick.Stop()
+				return
+			case <-tick.C:
+				c.rebuildRules()
+				c.runHandlers()
+			}
+		}
+	}()
+}
+
+// ConfigDescriptors supported by the CF Controller are only virtual-services
+// and destination-rules
+func (c *controller) ConfigDescriptor() model.ConfigDescriptor {
+	return model.ConfigDescriptor{
+		model.VirtualService,
+		model.DestinationRule,
+	}
+}
+
+// Get has been optimized for the current cloudfoundry case of
+// no namespace and only two specific types. We  have also
+// left these in maps for faster lookup
+func (c *controller) Get(typ, name, namespace string) (*model.Config, bool) {
+	c.storage.RLock()
+	defer c.storage.RUnlock()
+
+	var (
+		config *model.Config
+		found  bool
+	)
+
+	switch typ {
+	case model.VirtualService.Type:
+		if vs, ok := c.storage.virtualServices[name]; ok {
+			config = vs
+			found = true
+		}
+	case model.DestinationRule.Type:
+		if dr, ok := c.storage.destinationRules[name]; ok {
+			config = dr
+			found = true
+		}
+	default:
+		c.logger.Infof("get type not supported: %s", typ)
+	}
+
+	return config, found
+}
+
+// List also ignores namespace to return all the rules of a particular type
+// we've paid some attention here to avoid locking too long and also
+// avoid allocation performance hits
+func (c *controller) List(typ, namespace string) ([]model.Config, error) {
+	var configs []model.Config
+	var index int
+
+	c.storage.RLock()
```

Comment:
I'm a little surprised this is faster than the atomic value version, but I guess read locks are cheap.